home/.chezmoiscripts/universal/run_onchange_before_11-join-etcd-cluster.tmpl

```diff
@@ -0,0 +1,38 @@
+{{- if (eq .host.distro.family "linux") -}}
+#!/usr/bin/env bash
+# @file etcd Clustering
+# @brief Adds the host to `etcd` cluster
+# @description
+#     This script adds the host to `etcd` cluster provided as input, if `etcd` is installed. Static clustering
+#     mechanism is used to join the host to an existing etcd cluster. The other mechanisms that are based on
+#     service discovery is not supported by this script.
+#
+#     Static mechanism needs the IP address(es) of all the participating members to bootstrap a node/member.
+#     This is the only input needed to be able to add the host as a member of the cluster. The format of this field
+#     is available in the [inputs file](https://github.com/megabyte-labs/install.doctor/blob/master/home/.chezmoi.yaml.tmpl).
+#
+#     ## Configuration Variables
+#
+#     The following chart details the input variable(s) that are used to join the host to `etcd` cluster:
+#
+#     | Variable                     | Description                                                |
+#     |------------------------------|------------------------------------------------------------|
+#     | `etcd.initial_cluster_urls`  | Appropriately formatted initial cluster configuration      |
```

Comment:
Can we replace this with an address generated based on the domain address that gets populated in `.chezmoi.yaml.tmpl`? That way we can provision etcd even before we know the final IP address of the etcd server. Also, what is the `node` part for? Shouldn't we just need a server to connect to? Also, we should be using `https`.  And where did you get the two IP addresses from listed above?

I think we should run this over CloudFlare tunnels which I'll be setting up so you can assume that etcd.{{ service_domain }} points to the server. Just not sure what the node1 is for.

Reply:
The script is adding a member to an `etcd` cluster. If we are bootstrapping a new cluster, in which the host will be a member, then a discover mechanism can be used. In that case, the variable can be removed. This will let the cluster be created before the list of members and their addresses are known. This is different from adding a member to a cluster.

The `node` part is the name of the member in the cluster. I am not sure why there is a 'mapping' between the name and its advertise URLs. The value of `advertise-peer-urls` holds this 'map' of all the members that are part of the cluster and it changes when the membership changes. 

The IPs are just examples. I knew this would need changes, so I put it in to show what is expected.

Yes, we should be using HTTPS. Certificates can be passed as inputs when starting a member to encrypt communication between peers and with clients. It is also possible to have the members automatically generate certificates for communication among peers. Certificates can be used for authentication as well.

Reply:
Hey @enggnr -- we need to come up with a clever plan to use the PUBLIC_SERVICES_DOMAIN's DNS record on CloudFlare to figure out all the other peers, masters and then either provision the master node or join to it with a peer.

Each network provisioned by Install Doctor will use the PUBLIC_SERVICES_DOMAIN to classify which network to join to.